### PR TITLE
Added crosshairs to the secondary charts

### PR DESCRIPTION
**Pull request by rjmcneill**
Relates to #576

### CODE REVIEW

File: src/assets/js/chart/secondary/base.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import d3 from 'd3';
 import fc from 'd3fc';
 import event from '../../event';
+// import util from '../../util/util';
```
Comment (jdhodges) [inline src/assets/js/chart/secondary/base.js]: Remove comment

File: src/assets/js/chart/secondary/volume.js [hunk 2]
```diff
@@ -5,25 +5,75 @@ import event from '../../event';
...
+                crosshairDataPoint = updatedCrosshairData[0].datum;
+                dispatch.crosshairChange(updatedCrosshairData[0].datum);
+            } else {
+                crosshairDataPoint = undefined;
+                dispatch.crosshairChange(undefined);
+            }
+        })
+        .on('trackingend', function() {
+            dispatch.crosshairChange(undefined);
+        });
```
Comment (jdhodges) [inline src/assets/js/chart/secondary/volume.js]: Any way we can avoid repeating this crosshair creation code? Maybe a util function or separate component?

File: src/assets/js/model/chart/primary.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ import util from '../../util/util';
 export default function(initialProduct) {
     var model = {
         data: [],
+        currentCrosshairData: null,
```
Comment (jdhodges) [inline src/assets/js/model/chart/primary.js]: I'd be inclined to just call this `crosshairData`

File: src/assets/js/model/chart/secondary.js [hunk 4]
```diff
@@ -1,8 +1,27 @@
+import candlestickSeries from '../../series/candlestick';
+import option from '../menu/option';
+import util from '../../util/util';
+
 export default function(initialProduct) {
-    return {
+    var model = {
         data: [],
+        currentCrosshairData: null,
```
Comment (jdhodges) [inline src/assets/js/model/chart/secondary.js]: Same as previous comment.

File: src/assets/styles/primary.less [hunk 5]
```diff
@@ -62,34 +62,5 @@
...
-        > line {
-          stroke: @crosshair-colour;
-          stroke-dasharray: 0;
-        }
-      }
-    }
-  }
-
   background-color: @background-colour;
 }
```
Comment (jdhodges) [inline src/assets/styles/primary.less]: Need new line at the end of the file, must have been removed at some point.

File: src/assets/js/chart/secondary/macd.js [hunk 6]
```diff
@@ -1,37 +1,104 @@
...
+              .style('stroke-width', width);
+        });
+    }
+
+    var initialised = false;
+    var macdSeries;
+
+    function initialiseSeries(primarySeries) {
+        // Initialise the series to allow the secondaries to snap correctly
+        macdSeries = jquery.extend({}, primarySeries);
```
Comment (rjmcneill) [inline src/assets/js/chart/secondary/macd.js]: This is needed to ensure we can snap to data before the bar chart part of the macd indicator starts.

File: src/assets/js/app.js [hunk 7]
```diff
@@ -213,6 +217,8 @@ export default function() {
 
     function onCrosshairChange(dataPoint) {
         model.legend.data = dataPoint;
+        model.primaryChart.crosshairData = dataPoint;
+        model.secondaryChart.crosshairData = dataPoint;
```
Comment (jleft) [inline src/assets/js/app.js]: Are either of these used anywhere?

File: src/assets/js/util/crosshair.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import d3 from 'd3';
+import fc from 'd3fc';
+import event from '../event';
+import util from './util';
+
+export default function() {
```
Comment (jleft) [inline src/assets/js/util/crosshair.js]: I think it's a very good idea to separate this, but I don't think this is a util, I'd move this to a new `tool` namespace.

File: src/assets/js/model/chart/secondary.js [hunk 9]
```diff
@@ -1,9 +1,28 @@
...
         data: [],
+        crosshairData: null,
         viewDomain: [],
         trackingLatest: true,
         product: initialProduct,
         discontinuityProvider: initialDiscontinuityProvider
     };
+
+    var candlestick = candlestickSeries();
+    candlestick.id = util.uid();
```
Comment (jleft) [inline src/assets/js/model/chart/secondary.js]: It feels pretty odd having the primary chart series on the secondary charts model, but I can see why - let's discuss if there's an alternative.

If this was to stay, then until an indicator is changed, both the primary chart model and secondary chart model will be storing a reference to a different candlestick series, with a different ID. These should both be pulled up a level and passed into the models when we initialise them. (I'd leave this for now though until we discuss the above.)

File: src/assets/js/model/chart/secondary.js [hunk 10]
```diff
@@ -1,9 +1,28 @@
...
+    var _primarySeries = option('Candlestick', 'candlestick', candlestick);
+    _primarySeries.option.extentAccessor = ['high', 'low'];
+    Object.defineProperty(model, 'primarySeries', {
+        get: function() { return _primarySeries; },
+        set: function(newValue) {
+            _primarySeries = newValue;
+            model.selectorsChanged = true;
```
Comment (jleft) [inline src/assets/js/model/chart/secondary.js]: This is unused by the secondary charts and is not initialised anywhere, I don't think we need to use `Object.defineProperty`

File: src/assets/js/chart/secondary/macd.js [hunk 11]
```diff
@@ -1,37 +1,93 @@
 import d3 from 'd3';
```
Comment (jleft) [inline src/assets/js/chart/secondary/macd.js]: There's quite a lot of duplication between all of the secondary charts, I think some should be able to be pulled out into the base component - let's discuss further.

**Pull request closed without merging**